src/Provider/Settings/SettingsProvider.php

```diff
@@ -29,6 +30,11 @@ public function getSettings(IndexScope $indexScope): Settings
 
         $settings->synonyms = $this->synonymResolver->resolve($indexScope);
 
+        // todo this is incredibly ugly
```

Comment:
This is really ugly.

My idea for fixing it is to make a composite settings provider setup as we've done several times. The `ProductSettingsProvider` could then decorate the default settings provider and just add those ranking rules.